orion/primitives/tadgan.py

```diff
@@ -82,235 +127,254 @@ def __getstate__(self):
+
+        return output
+
+    @staticmethod
+    def _wasserstein_loss(y_true, y_pred):
+        return tf.reduce_mean(y_true * y_pred)
+
+    @staticmethod
+    def _gradient_penalty_loss_wrapper(critic):
+        def _gradient_penalty_loss(real, fake):
+            # Random weighted average to create interpolated signals.
+            batch_size = tf.shape(real)[0]
+            alpha = tf.random.uniform([batch_size, 1, 1], dtype=tf.float64)
+            interpolated = (alpha * real) + ((1 - alpha) * fake)
+
+            # Get the critic output for this interpolated signal.
+            with tf.GradientTape() as gp_tape:
+                gp_tape.watch(interpolated)
+                validity_interpolated = critic(interpolated)
+
+            # Calculate the gradients w.r.t to this interpolated signal.
+            grads = gp_tape.gradient(validity_interpolated, [interpolated])[0]
+
+            # Calculate the norm of the gradients.
+            norm = tf.sqrt(tf.reduce_sum(tf.square(grads), axis=np.arange(1, len(grads.shape))))
+            gradient_penalty = tf.reduce_mean((norm - 1.0) ** 2)
+            return gradient_penalty
+
+        return _gradient_penalty_loss
+
+    def _build_tadgan(self, **kwargs) -> None:
         hyperparameters = self.hyperparameters.copy()
         hyperparameters.update(kwargs)
 
-        self.encoder = self._build_model(hyperparameters, self.layers_encoder,
-                                         self.encoder_input_shape)
-        self.generator = self._build_model(hyperparameters, self.layers_generator,
-                                           self.generator_input_shape)
-        self.critic_x = self._build_model(hyperparameters, self.layers_critic_x,
-                                          self.critic_x_input_shape)
-        self.critic_z = self._build_model(hyperparameters, self.layers_critic_z,
-                                          self.critic_z_input_shape)
+        # Models
+        self.encoder = self._build_model(
+            hyperparameters, self.layers_encoder, self.encoder_input_shape, name='encoder')
+        self.generator = self._build_model(
+            hyperparameters, self.layers_generator, self.generator_input_shape, name='generator')
+        self.critic_x = self._build_model(
+            hyperparameters, self.layers_critic_x, self.critic_x_input_shape, name='critic_x')
+        self.critic_z = self._build_model(
+            hyperparameters, self.layers_critic_z, self.critic_z_input_shape, name='critic_z')
+
+        # Optimizers
+        self.critic_x_optimizer = import_object(self.optimizer)(self.learning_rate)
+        self.critic_z_optimizer = import_object(self.optimizer)(self.learning_rate)
+        self.encoder_generator_optimizer = import_object(self.optimizer)(self.learning_rate)
+
+        x = tf.keras.Input(shape=self.shape)
+        y = tf.keras.Input(shape=self.target_shape)
+        z = tf.keras.Input(shape=(self.latent_dim, 1))
```

Comment:
why not use `self.latent_shape` directly? 

There is a mismatch here between `(self.latent_dim, 1)` and `self.latent_shape` which is `(self.latent_dim, output_dim)`. I'm assuming that the latent shape is the correct dimension based on the architecture of critic z and the output of the encoder.